File: clojure/anagram/src/anagram.clj
(ns anagram
  (:require 
    [clojure.string :refer [lower-case]]))

(defn anagrams-for [word prospect-list]
  (let [word' (lower-case word)
        same-letters (comp (partial = (sort word')) sort)
        not-same-word (partial not= word')
        is-anagram #(->> %
                         lower-case
                         ((juxt same-letters not-same-word))
                         (every? true?))]
    (filter is-anagram prospect-list)))

File: scala/anagram/src/main/scala/Anagram.scala
object Anagram {
  private def sameLetters(a: String, b: String): Boolean =
    a.toLowerCase.sorted == b.toLowerCase.sorted

  private def sameWord(a: String, b: String): Boolean =
    a.toLowerCase == b.toLowerCase

  def findAnagrams(word: String, words: List[String]): List[String] =
    words.filter(x => !sameWord(word, x) && sameLetters(word, x))
}
